include/cuda/cudaexcept

```diff
@@ -0,0 +1,36 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_EXCEPTION
+#define _CUDA_EXCEPTION
+
+#include "std/detail/__config"
+#include "std/detail/__pragma_push"
+#include "std/version"
+#include "detail/cuda_util"
+
+#include <cuda_runtime_api.h>
+#include <stdexcept>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+/**
+ * @brief Exception thrown when a CUDA runtime error is encountered.
+ *
+ */
+struct cuda_error : ::std::runtime_error {
+  using ::std::runtime_error::runtime_error;
```

Comment:
I don't think this is right. Move the message string building here, and also add a way to query the error code that was encountered to this class (so that per-error-code handling of errors can be done when one of these is caught).

Reply:
Do you want just the raw `cudaError_t` error code? Or create a thing strongly typed wrapper around `cudaError_t`?

Reply:
See https://github.com/NVIDIA/libcudacxx/pull/105/commits/9703fb626a5904393f763fbc2107950932abe5d0